onnxruntime/core/session/inference_session.cc

```diff
@@ -422,6 +422,24 @@ common::Status InferenceSession::Initialize() {
                                                    std::make_unique<CPUExecutionProvider>(epi)));
     }
 
+    if (!session_options_.enable_sequential_execution) {
+      for (auto it = execution_providers_.begin(); it != execution_providers_.end(); ++it) {
+        // very rare case - user has registered CPU provider first -
+        // so parallel execution is still possible as all nodes will get assigned CPU provider
+        if (it->get()->Type() == onnxruntime::kCpuExecutionProvider) {
```

Comment:
Should we instead check for this scenario (CPU provider is registered first) and return an error at registration time as most likely the user is making a mistake, instead of adding special case handling for it here?

Reply:
That would also simplify the messaging I think as you wouldn't need the "as first priority" part for the CUDA error. What "as first priority" means isn't entirely clear as RegisterExecutionProvider doesn't mention 'priority' in its documentation. 

---
In reply to: [282268097](https://github.com/microsoft/onnxruntime/pull/988#discussion_r282268097) [](ancestors = 282268097)

Reply:
> Should we instead check for this scenario (CPU provider is registered first) and return an error at registration time as most likely the user is making a mistake, instead of adding special case handling for it here?

If I understand correctly, are you suggesting a change to not allow any provider to get registered after CPU Provider has already been registered? That assumes that the current node assignment logic will always be the same (i.e.) all nodes would get assigned to CPU Provider and anything registered after that is not considered at all. I would rather not introduce such a strong check in this PR and remove such a capability (some provider getting registered after the CPU provider) altogether even though it is implicit.

Reply:
I talked to @pranavsharma offline and if none of the execution providers are conducive to be used in parallel execution mode, just have a simple check to ensure that only the CPU Provider is registered in case parallel execution mode is desired. If that's not the case, return an appropriate error message.